### PR TITLE
fix(barrier): Fix bug about barrier collect 

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***
After https://github.com/singularity-data/risingwave/pull/4442 , If we use sync(epoch), we will sync all epochs less than or equal to this epoch to S3. Therefore, we need all epochs(le epoch) not to be written after we sync(epoch). 
However, if we drop the actors, the small epochs will flow through more actors than the large epochs.
This means that as we sync epoch, other epochs smaller than this epoch will be written.
It may be the reason for https://github.com/singularity-data/risingwave/issues/4774

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 1]
```diff
@@ -39,14 +39,14 @@ enum ManagedBarrierStateInner {
         remaining_actors: HashSet<ActorId>,
 
         /// Notify that the collection is finished.
-        collect_notifier: oneshot::Sender<CollectResult>,
+        collect_notifier: Option<oneshot::Sender<CollectResult>>,
```
Comment (hzxa21) [inline src/stream/src/task/barrier_manager/managed_state.rs:42]: This change is unneccessary?

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 2]
```diff
@@ -71,47 +73,58 @@ impl ManagedBarrierState {
...
+                .find_or_last(|(_, inner)| match inner {
+                    ManagedBarrierStateInner::Issued {
+                        remaining_actors, ..
+                    } => !remaining_actors.is_empty(),
+                    _ => true,
                 })
-                .collect();
+                .map(|(key, _)| *key);
+            let result = match result {
+                Some(epoch) => self.epoch_barrier_state_map.split_off(&epoch),
```
Comment (hzxa21) [inline src/stream/src/task/barrier_manager/managed_state.rs:87]: `split_off` returns everything after epoch but I think we want the entries before the epoch.

Comment (hzxa21) [reply src/stream/src/task/barrier_manager/managed_state.rs:87]: Can we add a unit test to cover the bug fixed in this PR?

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 3]
```diff
@@ -71,47 +73,58 @@ impl ManagedBarrierState {
...
-                        )
+                match inner {
+                    ManagedBarrierStateInner::Issued {
+                        collect_notifier, ..
+                    } => {
+                        // Notify about barrier finishing.
+                        let result = CollectResult {
+                            create_mview_progress,
+                        };
+                        // let collect_notifier = collect_notifier.take().unwrap();
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:114]:
```suggestion
```

**Comment by StrikeW** [comment]
> However, if we drop the actors, the small epochs will flow through more actors than the large epochs.

Could you clarify a bit for this statement? Why small epochs will flow through more actors than the large epochs?

**Comment by xxhZs** [comment]
> > However, if we drop the actors, the small epochs will flow through more actors than the large epochs.
> 
> Could you clarify a bit for this statement? Why small epochs will flow through more actors than the large epochs?

for example, We have actors (1 2 3), and we drop actor 2 3 with barrier1, barrier1 will flow through 1 2 3. But barrier2 will flow through 1. So barrier2 may be collected faster than barrier1.

File: src/storage/src/hummock/local_version_manager.rs [hunk 4]
```diff
@@ -395,6 +395,13 @@ impl LocalVersionManager {
         is_remote_batch: bool,
         table_id: u32,
     ) -> HummockResult<usize> {
+        let max_sync_epoch = self.local_version.read().get_max_sync_epoch();
+        assert!(
+            epoch > max_sync_epoch,
+            "write epoch must greater than sync epoch, write epoch{}, sync epoch{}",
+            epoch,
+            max_sync_epoch
+        );
```
Comment (hzxa21) [inline src/storage/src/hummock/local_version_manager.rs:404]: We can do the check in `write_shared_buffer_inner` to avoid acquiring the read lock.

**Comment by hzxa21** [comment]
@xxhZs Let's prioritize merging this PR first since it may fix the recent issues on parallel e2e test failure.

**Review by hzxa21 — APPROVED** [review]
Rest LGTM.

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 5]
```diff
@@ -71,47 +71,54 @@ impl ManagedBarrierState {
...
-                _ => unreachable!(),
             }
         }
     }
 
     /// Remove stop barrier (epoch < `curr_epoch`), and send err.
     pub(crate) fn remove_stop_barrier(&mut self, curr_epoch: u64) {
-        self.epoch_barrier_state_map
-            .drain_filter(|k, _| k < &curr_epoch);
+        self.epoch_barrier_state_map.retain(|k, _| k > &curr_epoch);
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:121]: Should be `>=` here? cc @yezizp2012

Comment (yezizp2012) [reply src/stream/src/task/barrier_manager/managed_state.rs:121]: Yes, it should be `>=` here.

**Comment by BugenZhao** [comment]
I recall that we also have a similar logic on the meta service for sequentially committing. But that one is for global ordering while this is for compute nodes. Am I correct here?

https://github.com/risingwavelabs/risingwave/blob/3af5280fd76bc2fdcc3cb585bd0d865ef53a5247/src/meta/src/barrier/mod.rs#L780-L819

One concern for me is that we cannot distinguish between the barriers with the state "not collected" and "collected but not able to commit due to ordering" on the await-tree (async stack trace) of compute nodes. 🤔

**Comment by xxhZs** [comment]
> I recall that we also have a similar logic on the meta service for sequentially committing. But that one is for global ordering while this is for compute nodes, so we need both of them. Am I correct here?
> 
> https://github.com/risingwavelabs/risingwave/blob/3af5280fd76bc2fdcc3cb585bd0d865ef53a5247/src/meta/src/barrier/mod.rs#L780-L819
> 
> One concern for me is that we cannot distinguish between the barriers with the state "not collected" and "collected but not able to commit due to ordering" on the await-tree (async stack trace) of compute nodes. 🤔

yes, It is. We have similar logic in the meta and computer nodes to guarantee different things.
If the code has not been changed, We cannot distinguish whether the state of barries in `ManagedBarrierState`'s queue is "not collected"  or "collected but not able to commit due to ordering" . But "collected but not able to commit due to ordering" will only appear when we drop actor

**Pull request merged**